Add Avri waste collection sensor

timvancann opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
Adds Dutch Avri waste collecting sensor


## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
```yaml
sensor:
  - platform: avri
    postcode: 1234AB
    house_number: 42
```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]
https://github.com/home-assistant/home-assistant.io/pull/11867

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

timvancann committed:
Add Avri waste collection sensor

timvancann committed:
Apply black formatting

timvancann committed:
Update manifest

timvancann committed:
Add requirements

timvancann committed:
Add sensor to coverage

timvancann committed:
Update import order

springstan commented:
@timvancann please link the documentation PR :)

timvancann commented:
@springstan done!

springstan commented on homeassistant/components/avri/sensor.py:
```diff
@@ -0,0 +1,100 @@
+    {
+        vol.Required(CONF_POSTCODE): cv.string,
+        vol.Required(CONF_HOUSE_NUMBER): cv.string,
+        vol.Optional(CONF_HOUSE_NUMBER_EXTENSION, default=""): cv.string,
```
Please try to avoid empty strings as default values. Does this house number extension need a default value at all?

timvancann replied:
Fixed it by moving it to the constructor.

springstan commented on homeassistant/components/avri/sensor.py:
```diff
@@ -0,0 +1,100 @@
+                    self._state = event.day.date()
+                    break
+            else:
+                self._state = None
```
In which case is this else statement entered? No if statement is preceeding it so I am not sure when that would be the case.

timvancann replied:
Good point, fixed

timvancann committed:
Bump dependency to include todays pickup

timvancann committed:
Bump avri version in requirements_all.txt

timvancann commented on homeassistant/components/avri/sensor.py:
```diff
@@ -0,0 +1,100 @@
+            else:
+                self._state = None
+        except AvriException as ex:
+            _LOGGER.error("Avri platform error. %s", ex)
```
I logged the exception as it contains a proper message.

timvancann commented on homeassistant/components/avri/sensor.py:
```diff
@@ -0,0 +1,100 @@
+
+    def update(self):
+        """Update device state."""
+        try:
```
Done.

timvancann commented on homeassistant/components/avri/sensor.py:
```diff
@@ -0,0 +1,100 @@
+    else:
+        for upcoming in each_upcoming:
+            add_entities(
+                [AvriWasteUpcoming(config[CONF_NAME], client, upcoming.name)], True
```
Good point! Fixed

timvancann committed:
Code review comments

springstan commented on homeassistant/components/avri/sensor.py:
```diff
@@ -0,0 +1,102 @@
+    client = Avri(
+        postal_code=config[CONF_POSTCODE],
+        house_nr=config[CONF_HOUSE_NUMBER],
+        house_nr_extension=config.get(CONF_HOUSE_NUMBER_EXTENSION, ""),
```
Can you change it inside of your library to handle or accept `None` as the default value for the house_number_extension?

springstan commented on homeassistant/components/avri/sensor.py:
```diff
@@ -0,0 +1,102 @@
+    try:
+        each_upcoming = client.upcoming_of_each()
+    except AvriException as ex:
+        raise PlatformNotReady from ex
```
```suggestion
        raise PlatformNotReady
```

timvancann replied:
Can you explain why? The reason for the platform not being ready is directly caused by some exception in the API. The `from` seems valid here.

springstan replied:
Oh yeah sorry you are correct.

springstan commented on homeassistant/components/avri/sensor.py:
```diff
@@ -0,0 +1,102 @@
+        vol.Required(CONF_POSTCODE): cv.string,
+        vol.Required(CONF_HOUSE_NUMBER): cv.string,
+        vol.Optional(CONF_HOUSE_NUMBER_EXTENSION): cv.string,
+        vol.Optional(CONF_COUNTRY_CODE, default="NL"): cv.string,
```
Create a constant for the default country code as well.

timvancann replied:
Done.

springstan commented on homeassistant/components/avri/sensor.py:
```diff
@@ -0,0 +1,102 @@
+        try:
+            pickup_events = self.client.upcoming_of_each()
+        except AvriException as ex:
+            _LOGGER.error(ex)
```
Explain where and maybe why an error is occurring.

timvancann replied:
The exception message explains the why quite well, I did add the where

timvancann committed:
Reduce scan interval to 4 hours

This makes sure that no matter what happens, in
the morning the correct dates have been pulled
without the old ones lingering for too long.

timvancann committed:
Better logging

springstan commented on homeassistant/components/avri/sensor.py:
```diff
@@ -0,0 +1,105 @@
+CONF_HOUSE_NUMBER_EXTENSION = "house_number_extension"
+DEFAULT_NAME = "avri"
+ICON = "mdi:trash-can-outline"
+SCAN_INTERVAL = 14_400
```
Not sure if the scan_interval can contain underscores.
```suggestion
SCAN_INTERVAL = 14400
```

timvancann replied:
Since PEP 515 (https://www.python.org/dev/peps/pep-0515/) any numeric literal can contain underscores.

However, to avoid confusion and improve readability I made it a `timedelta` as other components also seem to use that :).

springstan replied:
Thanks did not know that :)

That is even better 👍 FYI you will need to run `isort` to fix the pylint error

timvancann committed:
Made scan interval a timedelta

timvancann committed:
Fix import order

springstan commented on homeassistant/components/avri/sensor.py:
```diff
@@ -0,0 +1,106 @@
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
+    {
+        vol.Required(CONF_POSTCODE): cv.string,
+        vol.Required(CONF_HOUSE_NUMBER): cv.string,
```
Wouldn't it make more sense to use `int` in this case? Or are there house numbers that cannot be displayed in an `int`?

Furthermore, the example you give for the library is also using int:

```python
from avri.api import Avri
client = Avri('1234AB', 42)
```

```suggestion
        vol.Required(CONF_HOUSE_NUMBER): cv.positive_int,
```

timvancann replied:
Good point! Didn't know there was a `positive_int` type. nice one.

springstan replied:
Did you intend to revert this with your latest commit?

timvancann replied:
Ah, I did not. Think something went wrong whilst rebasing. Fixed it.

springstan replied:
Now I am uncertain whether or not the `house_number`, `house_number_extension` or both can only have a positive int as their value 😅

timvancann replied:
Ah dang it... I meant to change `house_number`. Extension can be anything. I'm so sorry for the confusion. Fixed it for real this time!

springstan replied:
Thanks for the clarification!

timvancann committed:
Update homeassistant/components/avri/sensor.py

Co-Authored-By: springstan <46536646+springstan@users.noreply.github.com>

springstan commented on homeassistant/components/avri/sensor.py:
```diff
@@ -0,0 +1,106 @@
+            )
+            self._state = None
+        else:
+            if len(pickup_events) == 0:
```
Use the empty list as a falsy value:
```suggestion
            if not pickup_events:
```

springstan commented on homeassistant/components/avri/sensor.py:
```diff
@@ -0,0 +1,106 @@
+            for event in pickup_events:
+                if event.name == self._waste_type:
+                    self._state = event.day.date()
+                    break
```
What do you think about using the filter function in combination with a lambda function to avoid the break statement?
Something along the lines of:

```python
filtered_event = list(filter(lambda event: event.name is self._waste_type))
self._state = filtered_event.day.date()
```

timvancann replied:
Nice idea! I made some changes that would streamline it a bit more.

timvancann committed:
Update homeassistant/components/avri/sensor.py

Co-Authored-By: springstan <46536646+springstan@users.noreply.github.com>

timvancann committed:
Use filter instead of break statement

timvancann committed:
Use positive int for house number extension

timvancann committed:
Switch voluptuous types for house number and house number extension

springstan reviewed (commented):
Looks great @timvancann 👍 just one small comment:

springstan commented on homeassistant/components/avri/sensor.py:
```diff
@@ -0,0 +1,106 @@
+            if not matched_events:
+                self._state = None
+                return
+            self._state = matched_events[0].day.date()
```
In which format is this date returned?

timvancann replied:
Default string representation of a python `date` format: `YYYY-MM-DD`

springstan commented on homeassistant/components/avri/sensor.py:
```diff
@@ -0,0 +1,106 @@
+class AvriWasteUpcoming(Entity):
+    """Avri Waste Sensor."""
+
+    def __init__(self, name: str, client: Avri, waste_type: str):
```
```suggestion
    def __init__(self, name, client, waste_type):
```

timvancann replied:
I absolutely love this python3.5+ typehints introduced in https://www.python.org/dev/peps/pep-0484/. It helps the readers of the code, and also any modern editor with developing.

Is there a particular reason they should be removed?

springstan replied:
Well I just wanted it to be consistent with other methods in that class but I think it is fine if they stay in this case.

springstan reviewed (approved):
LGTM! ✌ just one more suggestion, btw please update the documentation :)

balloob commented on homeassistant/components/avri/sensor.py:
```diff
@@ -0,0 +1,106 @@
+    client = Avri(
+        postal_code=config[CONF_POSTCODE],
+        house_nr=config[CONF_HOUSE_NUMBER],
+        house_nr_extension=config[CONF_HOUSE_NUMBER_EXTENSION],
```
This is an optional value, and so can be omitted. Calling it with `[]` will raise a `KeyError`.
```suggestion
        house_nr_extension=config.get(CONF_HOUSE_NUMBER_EXTENSION),
```

springstan replied:
Good catch! Missed that one 😄

balloob commented on homeassistant/components/avri/sensor.py:
```diff
@@ -0,0 +1,106 @@
+    def unique_id(self) -> str:
+        """Return a unique ID."""
+        return (
+            f"{self.name}"
```
name should not be part of the unique id.

timvancann replied:
How about 
```
            f"avri_{self._waste_type}"
            f"{self.client.country_code}{self.client.postal_code}"
            f"{self.client.house_nr}{self.client.house_nr_extension}"
```
?

balloob replied:
makes sense. You can drop `avri_`, it is checked only within your integration sensor platform.

springstan replied:
This still needs to be changed :)

balloob commented on homeassistant/components/avri/sensor.py:
```diff
@@ -0,0 +1,106 @@
+            _LOGGER.error(
+                "There was an error retrieving upcoming garbage pickups: %s", ex
+            )
+            self._state = None
```
You also would want to set `available` to return `False`

timvancann replied:
done!

balloob commented:
Also, the link to GitHub on your PyPI lib links to sampleproject

springstan commented on homeassistant/components/avri/sensor.py:
```diff
@@ -0,0 +1,106 @@
+                filter(lambda event: event.name == self._waste_type, pickup_events)
+            )
+            if not matched_events:
+                self._state = None
```
If `available` is completely implemented it should be set to `False` here as well

timvancann replied:
done!

timvancann committed:
Update homeassistant/components/avri/sensor.py

Co-Authored-By: Paulus Schoutsen <paulus@home-assistant.io>

timvancann committed:
Implement `available`

timvancann committed:
Bump avri api

timvancann commented:
@balloob very good catch, I missed that :sweat_smile: . I update the link on PyPi.

springstan commented on homeassistant/components/avri/sensor.py:
```diff
@@ -0,0 +1,111 @@
+            _LOGGER.error(
+                "There was an error retrieving upcoming garbage pickups: %s", ex
+            )
+            self._state = False
```
The state should not be `False` for a sensor at least it does not make sense in this case. Set the state to `None` here since `bool(None)` results in `False` for the available property.
```suggestion
            self._state = None
```

balloob commented on homeassistant/components/avri/sensor.py:
```diff
@@ -0,0 +1,111 @@
+            matched_events = list(
+                filter(lambda event: event.name == self._waste_type, pickup_events)
+            )
+            if not matched_events:
```
So right now it means that if you can't find any events, you're going to mark it as unavailable. But that means you're treating 0 upcoming events the same as if you cannot access the API.

springstan replied:
I am not sure but should not there always be an upcoming event? Even if it is far into the future it is still an upcoming event.

timvancann replied:
@balloob Your assumption is correct. How would use suggest the state of the sensor should be if there's no upcoming event?

@springstan I try not to make assumptions about any external api. Even though `0` might seem unlikely, I'd rather account for it than run into issues. The external api is weird enough as it is :p.

springstan replied:
Fair enough 😄 what do you think about using `STATE_UNKNOWN` for this case?
https://github.com/home-assistant/home-assistant/blob/21f8d0a80c788f01aec418a2a7d712275cca3f72/homeassistant/const.py#L204

Reflecting that even the api does not know when the next upcoming event is. Though I would recommend to add this to the documentation brief users.

timvancann replied:
@springstan That's a nice one! I'll update the code and the docs accordingly.

springstan commented on homeassistant/components/avri/sensor.py:
```diff
@@ -0,0 +1,111 @@
+                filter(lambda event: event.name == self._waste_type, pickup_events)
+            )
+            if not matched_events:
+                self._state = False
```
Same here.
```suggestion
                self._state = None
```

springstan commented on homeassistant/components/avri/sensor.py:
```diff
@@ -0,0 +1,111 @@
+        entities = [
+            AvriWasteUpcoming(config[CONF_NAME], client, upcoming.name)
+            for upcoming in each_upcoming
+        ]
```
Please use the preferred way for for loops:
```python
for upcoming in each_upcoming
    entities.append(...)
```

balloob replied:
I think that using a list comprehension is fine.

springstan replied:
Okay sure.

timvancann committed:
Code review comments

timvancann committed:
Replace `postcode` with `zip_code`

balloob commented on homeassistant/components/avri/sensor.py:
```diff
@@ -0,0 +1,111 @@
+import voluptuous as vol
+
+from homeassistant.components.sensor import PLATFORM_SCHEMA
+from homeassistant.const import CONF_NAME, STATE_UNKNOWN
```
There is no reason to use `STATE_UNKNOWN`. Return `None` from the state property and the system will mark it as unknkown.

springstan replied:
Oh yeah right 😅 completely forgot about that. When is `STATE_UNKNOWN` used then?

balloob replied:
It is used internally when you return `None`. But it's better to let the system swap it out, in case the system ever wants to do other things when a state is unknown (currently it does not)

springstan replied:
Thanks for the quick answer! Just wanted to know what exactly happens behind the scenes with this particular state.

balloob commented on homeassistant/components/avri/sensor.py:
```diff
@@ -0,0 +1,111 @@
+    def unique_id(self) -> str:
+        """Return a unique ID."""
+        return (
+            f"avri_{self._waste_type}"
```
Remove `avri_`

timvancann replied:
Done.

balloob commented on homeassistant/components/avri/sensor.py:
```diff
@@ -0,0 +1,111 @@
+        """Return a unique ID."""
+        return (
+            f"avri_{self._waste_type}"
+            f"{self.client.country_code}{self.client.postal_code}"
```
You will want to add delimiters between the values.

timvancann replied:
Done.

balloob commented on homeassistant/components/avri/sensor.py:
```diff
@@ -0,0 +1,111 @@
+    @property
+    def available(self):
+        """Return True if entity is available."""
+        return bool(self._state)
```
This won't work anymore. You need to keep track of a separate value for `available` so you don't conflate unknown state with unable to access the API.

timvancann replied:
Does the `available` need to return `True` if the state is `"unknown"`?

balloob replied:
Yeah. `available` indicates if you can reach the API. `unknown` means that the API did not provide useful information.

balloob replied:
Technically it is known, in that there is nothing scheduled.

timvancann replied:
Got it, thanks!

timvancann committed:
Update logic for `available`

springstan commented on homeassistant/components/avri/sensor.py:
```diff
@@ -20,6 +20,7 @@
 ICON = "mdi:trash-can-outline"
 SCAN_INTERVAL = timedelta(hours=4)
 DEFAULT_COUNTRY_CODE = "NL"
+UNIQUE_ID_DELIMITER = "_"
```
A constant is not really necessary for a unique id, see the example in the [docs](https://developers.home-assistant.io/docs/entity_registry_index/#unique-id-requirements).

timvancann replied:
:+1:

timvancann committed:
Remove variable for delimiter

MartinHjelmare commented on homeassistant/components/avri/sensor.py:
```diff
@@ -0,0 +1,116 @@
+        else:
+            self._state_available = True
+            matched_events = list(
+                filter(lambda event: event.name == self._waste_type, pickup_events)
```
Just for reviewers for the future: We prefer list comprehension over filter + lambda.

Otherwise looks great!

springstan replied:
Hehe my bad 😅 thought that it was shorter and easier to read, but if list comprehension is preferred I will suggest it in the future.
